libyul/optimiser/UnusedStoreEliminator.cpp

```diff
@@ -0,0 +1,463 @@
+		sideEffects = builtin->controlFlowSideEffects;
+	else
+		sideEffects = m_controlFlowSideEffects.at(_functionCall.functionName.name);
+
+	if (!sideEffects.canContinue)
+	{
+		changeUndecidedTo(State::Unused, Location::Memory);
+		changeUndecidedTo(sideEffects.canTerminate ? State::Used : State::Unused, Location::Storage);
+	}
+}
+
+void UnusedStoreEliminator::operator()(FunctionDefinition const& _functionDefinition)
+{
+	ScopedSaveAndRestore storeOperations(m_storeOperations, {});
+	UnusedStoreBase::operator()(_functionDefinition);
+}
+
+
+void UnusedStoreEliminator::operator()(Leave const&)
+{
+	changeUndecidedTo(State::Used);
+}
+
+void UnusedStoreEliminator::visit(Statement const& _statement)
+{
+	using evmasm::Instruction;
+
+	UnusedStoreBase::visit(_statement);
+
+	auto const* exprStatement = get_if<ExpressionStatement>(&_statement);
+	if (!exprStatement)
+		return;
+
+	FunctionCall const* funCall = get_if<FunctionCall>(&exprStatement->expression);
+	if (!funCall)
+		return;
+	optional<Instruction> instruction = toEVMInstruction(m_dialect, funCall->functionName.name);
+	if (!instruction)
+		return;
+
+	if (!ranges::all_of(funCall->arguments, [](Expression const& _expr) -> bool {
+		return get_if<Identifier>(&_expr) || get_if<Literal>(&_expr);
+	}))
+		return;
+
+	if (
+		*instruction == Instruction::SSTORE ||
+		(!m_ignoreMemory && (
+			*instruction == Instruction::EXTCODECOPY ||
+			*instruction == Instruction::CODECOPY ||
+			*instruction == Instruction::CALLDATACOPY ||
+			*instruction == Instruction::RETURNDATACOPY ||
+			*instruction == Instruction::MSTORE ||
+			*instruction == Instruction::MSTORE8
+		))
+	)
+	{
+		m_stores[YulString{}].insert({&_statement, State::Undecided});
+		vector<Operation> operations = operationsFromFunctionCall(*funCall);
+		yulAssert(operations.size() == 1, "");
+		m_storeOperations[&_statement] = move(operations.front());
+	}
+}
+
+void UnusedStoreEliminator::finalizeFunctionDefinition(FunctionDefinition const&)
+{
+	changeUndecidedTo(State::Used);
+	scheduleUnusedForDeletion();
+}
+
+vector<UnusedStoreEliminator::Operation> UnusedStoreEliminator::operationsFromFunctionCall(
+	FunctionCall const& _functionCall
+) const
+{
+	using evmasm::Instruction;
+
+	YulString functionName = _functionCall.functionName.name;
+	SideEffects sideEffects;
+	if (BuiltinFunction const* f = m_dialect.builtin(functionName))
+		sideEffects = f->sideEffects;
+	else
+		sideEffects = m_functionSideEffects.at(functionName);
+
+	optional<Instruction> instruction = toEVMInstruction(m_dialect, functionName);
+	if (!instruction)
+	{
+		vector<Operation> result;
+		// Unknown read is worse than unknown write.
+		if (sideEffects.memory != SideEffects::Effect::None)
+			result.emplace_back(Operation{Location::Memory, Effect::Read, {}, {}});
+		if (sideEffects.storage != SideEffects::Effect::None)
+			result.emplace_back(Operation{Location::Storage, Effect::Read, {}, {}});
+		return result;
+	}
+
+	switch (*instruction)
```

Comment:
This should be pulled into into SemanticInformation.h or similar.

Of course we cannot work with yul strings here, but we could say:
 - this reads from and writes to memory
 - the read start can be found in argument 2
 - the read length is always 32
 - the write length can be found in argument 7

and so on.